### DISCUSSION TITLE
General libp2p network implementation questions.

### DISCUSSION THREAD

**Discussion by haider-rs** [libp2p/rust-libp2p#2779] [Q&A] [answered]
I want to use libp2p for building my p2p network where i can communicate with other nodes and transfer some struct between them.
i want to discuss some research done on libp2p and verify if collected data is correct and also got some questions.

Q:1
Are these protocols correct to be used for this scenario? or am i missing something or any of these is just a overhead?

Gossipsub: for events subscription between nodes and publishing events
mDNS: for looking peers into my local network
Kademlia: for making a DHT tree for peers can store information about each other.
Identify: to get peer information

Q:2
How can i get the address which we have to use in bootstrap nodes? since we create a identity at runtime while running a node.

Q:3
Is my understanding correct. i have will bootstrap nodes in my network and after implementing above protocols it will connect to bootstrap nodes and form a network and then i can use my gossipsub protocol to communicate?

Q:4 
Where does the holepunching implementation go? do we need it? when we want to connect a peer after getting information from indentify protocol?

looking for kind answers.

Thank you

**Comment by mxinden** [comment, marked as answer]
:wave: 

>  Are these protocols correct to be used for this scenario? or am i missing something or any of these is just a overhead?

Correct

>  How can i get the address which we have to use in bootstrap nodes? since we create a identity at runtime while running a node.

I am not sure I follow. Can you rephrase?

>  Is my understanding correct. i have will bootstrap nodes in my network and after implementing above protocols it will connect to bootstrap nodes and form a network and then i can use my gossipsub protocol to communicate?

Yep. Though make sure to connect to other nodes that you discover along the way. Otherwise you have a star topology where everyone is only connected to the boot node.

>  Where does the holepunching implementation go? do we need it? when we want to connect a peer after getting information from indentify protocol?

You need to integrate them manually at the moment, see hole punching tutorial.

https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html

Hope these answers are of some help.

**Reply by haider-rs** [reply]
thanks max!, and i was asking about the peer_id which we need to add in a array of bootstrap nodes.
Since we create identity (Peer ID) of each node at runtime! i need to add bootstrap node peer id in code before making a final executable. so if I check peer id after running executable how will I update the code! that's why I thought there might be a way to get peer id of node before make the final executable to I can make final executable with vec of peerIDs => bootstrap nodes. i hope I make sense.

**Reply by thomaseizinger** [reply]
You can serialise the generated `Keypair` and create one again from the serialised format. For example, the ed25519 variant has functions for going from and to bytes.

**Reply by mxinden** [reply]
@Haider-Ali-DS also take a look at the `keygen` tool: https://github.com/libp2p/rust-libp2p/tree/master/misc/keygen
